confdgnmi/src/confd_gnmi_adapter.py

```diff
@@ -58,126 +58,238 @@ def set(self, prefix, path, val):
+
+        def changes(self):
             """
-            Read (get) subscription response(s) (in stream) for added subscription requests
-            This should be generator function
-            Response contains `notification` or `sync_response`
-            :param streaming: indicates this is streaming read request (POLL, STREAM),
-                            if yes, keep processing and wait for SubscriptionStreamEventType.FINISH
-            :return: TODO
+            Get subscription response for changes (subscribed values).
+            `update` array contains changes
+            TODO `delete` is processed and `delete` array is empty
+            TODO `alias` is dummy string, atomic is always False
+            TODO timestamp is 0
+            :return: SubscribeResponse with changes
+            """
+            log.debug("==>")
+            update = self.get_monitored_changes()
+            notif = gnmi_pb2.Notification(timestamp=0,
+                                          prefix=self.subscription_list.prefix,
+                                          alias="/alias", update=update,
+                                          delete=[],
+                                          atomic=False)
+            response = gnmi_pb2.SubscribeResponse(update=notif)
+            log.debug("<== response=%s", response)
+            return response
+
+        def read(self):
+            """
+            Read (get) subscription response(s) (in stream) for added
+            subscription requests.
+            This is generator function. For streaming subscription it contains
+            event loop driven by self.read_queue (Queue object) and
+            SubscriptionEvent messages.
+            Response contains `notification` or `sync_response`.
+            :return: nothing
+            #TODO POLL mode with updates_only (send sync_response)
+            :see: SubscriptionHandler.SubscriptionEvent
             """
             log.info("==>")
             # TODO exceptions
-            self.subscription_stream_event_type = self.SubscriptionStreamEventType.GET
-            assert self.subscription_list != None
-            if streaming:
-                self.thread_event = Event()
+            assert self.subscription_list is not None
+            if not self.is_once():
+                self.read_queue = Queue()
+            event = None
+            first_sample = True
             while True:
-                if self.subscription_stream_event_type == self.SubscriptionStreamEventType.GET:
-                    log.debug("processing response")
-                    response = self.make_subscription_response()
+                log.debug("Processing event type %s", event)
+                # SAMPLE is handled in the same way as "first_sample"
+                if first_sample or event == self.SubscriptionEvent.SAMPLE:
+                    response = self.sample(
+                        start_monitoring=self.is_monitor_changes() and first_sample)
```

Comment:
I am sure this `while True` loop can be rewritten to a more Pythonic for loop, iterating e.g. over the read_queue; what I'm not sure is that it would be shorter or easier to read - did you have a look?

Reply:
No, this would mean implementing iterable. One option could be use Queue also for ONCE subscription, but I do not like instantiating Queue just for this.

Reply:
no changes